src/realm/table.cpp

```diff
@@ -1353,47 +1272,71 @@ void Table::migrate_objects(util::FunctionRef<void()> commit_and_continue)
 
     /******************** Optionally create !OID accessor ********************/
 
+    ColKey oid_col;
     BPlusTree<Int>* oid_column = nullptr;
     std::unique_ptr<BPlusTreeBase> oid_column_store;
     if (m_spec.get_column_name(0) == "!OID") {
-        // The !OID column is needed in the next stage as well, so it must be
-        // deleted in this stage, so move it from column_accessors to oid_column_store.
-        auto col0 = m_spec.get_key(0);
-        oid_column_store = std::move(column_accessors[col0]);
+        // The !OID column should not be migrated, but we need an accessor
+        // to the old column
+        oid_col = m_spec.get_key(0);
+        oid_column_store = std::move(column_accessors[oid_col]);
 
-        column_accessors.erase(col0);
+        column_accessors.erase(oid_col);
         oid_column = dynamic_cast<BPlusTree<Int>*>(oid_column_store.get());
         REALM_ASSERT(oid_column);
-
-        // Delete '0' from cols_to_destroy
-        auto first = cols_to_destroy.begin();
-        REALM_ASSERT(*first == 0);
-        cols_to_destroy.erase(first);
     }
 
     /*************************** Create objects ******************************/
 
+    bool use_row_ndx_as_key = !(oid_column || (pk_col_key && pk_col_key.get_type() == col_type_String));
```

Comment:
a bit too many negations involved in relation to this var (used negated below)

Reply:
perhaps negate and call it 'use_oid_or_pk' ?

Reply:
Yes, I was aware of that. I just liked the `use_row_ndx_as_key` name better. Simpler to understand than the negated version. And it is used both directly and negated afterwards.  I think `if not use row ndx as key` makes sense when reading it.